File: src/main/golang/leetcode/editor/cn/[300]Longest Increasing Subsequence.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func lengthOfLIS(nums []int) (ans int) {
	//Max := func(x, y int) int {
	//	if x > y {
	//		return x
	//	}
	//	return y
	//}
	//n := len(nums)
	//dp := make([]int, n)
	//for i := 0; i < n; i++ {
	//	dp[i] = 1
	//	for j := 0; j < i; j++ {
	//		if nums[i] > nums[j] {
	//			dp[i] = Max(dp[i], dp[j]+1)
	//		}
	//	}
	//	ans = Max(ans, dp[i])
	//}
	//return
	n := len(nums)
	top := make([]int, n)
	var piles int
	for _, num := range nums {
		l, r := 0, piles
		for l < r {
			mid := l + (r-l)/2
			if top[mid] < num {
				l = mid + 1
			} else {
				r = mid
			}
		}
		if l == piles {
			piles++
		}
		top[l] = num
	}
	return piles
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[300]Longest Increasing Subsequence.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # n = len(nums)
        # if n < 2: return n
        # dp = [1] * n
        # for i in range(1, n):
        #     for j in range(i):
        #         if nums[j] >= nums[i]: continue
        #         dp[i] = max(dp[i], dp[j] + 1)
        # return max(dp)
        piles, n = 0, len(nums)
        top = [0] * n
        for i in range(n):
            poker = nums[i]
            left, right = 0, piles
            ## 找到>=poker的最左边的堆
            while left < right:
                mid = (left + right) // 2
                if top[mid] < poker:
                    left = mid + 1
                else:
                    right = mid
            if left == piles:
                piles += 1
            top[left] = poker
        return piles
# leetcode submit region end(Prohibit modification and deletion)
